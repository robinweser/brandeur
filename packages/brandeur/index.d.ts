declare module 'brandeur' {
  import { CSSProperties } from 'react'
  import { createHooks as createBaseHooks } from '@css-hooks/react'
  import { WithHooks } from '@css-hooks/core'

  export type Style = CSSProperties
  export type PluginProperties<T> = T extends (_: infer Arg) => unknown
    ? Arg
    : Style

  export type Fallback = {
    property: string | Array<string>
    values: Array<string>
    match?: string
  }

  type HookOptions<Hooks extends string> = Parameters<
    typeof createBaseHooks<Hooks>
  >[0]

  export type Config<Hooks extends string, Theme, Plugins> = {
    hooks: HookOptions<Hooks>
    config?: Object
    theme?: Theme
    plugins?: Plugins
    fallbacks?: Array<Fallback>
  }

  export type Properties<Hooks, Plugins = []> = WithHooks<
    Hooks,
    Style | PluginProperties<Plugins[number]>
  >

  export type Rule<Theme, Props> = (theme: Theme) => Props
  export type Rules<Theme, Props> =
    | undefined
    | Props
    | Rule<Theme, Props>
    | Array<Rules<Theme, Props>>

  type CSSFunction<Theme, Props> = (
    properties_0: Rules<Theme, Props>,
    ...properties_1: (Rules<Theme, Props> | undefined)[]
  ) => CSSProperties

  export function createHooks<Hooks extends string, Theme = {}, Plugins = []>(
    config: Config<Hooks, Theme, Plugins>
  ): [string, CSSFunction<Theme, Properties<Hooks, Plugins>>]
}

declare module 'fela-plugin-bidi' {
  import { Style } from 'brandeur'

  export default function bidi(
    direction?: 'ltr' | 'rtl'
  ): (style: Style) => Style
}

declare module 'fela-plugin-custom-property' {
  import { Style } from 'brandeur'

  export default function customProperty<P extends Record<string, any>>(
    properties: P
  ): (
    style: Partial<{
      [Property in keyof P]: Parameters<P[Property]>[0]
    }>
  ) => Style
}

declare module 'fela-plugin-expand-shorthand' {
  import { Style } from 'brandeur'

  export default function expandShorthand(): (style: Style) => Style
}

declare module 'fela-plugin-extend' {
  import { Style } from 'brandeur'
  import { WithHooks } from '@css-hooks/core'

  type Extension<T, Hooks> = {
    condition: boolean
    style: ExtendStyle<T, Hooks>
  }

  type ExtendStyle<T, Hooks> = WithHooks<Hooks, T> & {
    extend?: Extension<T, Hooks> | Array<Extension<T, Hooks>>
  }

  export default function extend<Hooks, T = Style>(): (
    style: ExtendStyle<T, keyof Hooks>
  ) => Style
}

declare module 'fela-plugin-hover-media' {
  import { Style } from 'brandeur'

  export default function hoverMedia(): (style: Style) => Style
}

declare module 'fela-plugin-important' {
  import { Style } from 'brandeur'

  export default function important(): (style: Style) => Style
}

declare module 'fela-plugin-isolation' {
  import { Style } from 'brandeur'

  type Config = {
    exclude: Array<keyof Style>
  }
  export default function isolation(config?: Config): (style: Style) => Style
}

declare module 'fela-plugin-kebab-case' {
  import { Style } from 'brandeur'

  export default function kebabCase(): (style: Style) => Style
}

declare module 'fela-plugin-logger' {
  import { Style } from 'brandeur'

  export default function logger(): (style: Style) => Style
}

declare module 'fela-plugin-rtl' {
  import { Style } from 'brandeur'

  export default function rtl(
    direction?: 'ltr' | 'rtl'
  ): (style: Style) => Style
}

declare module 'fela-plugin-unit' {
  import { Style } from 'brandeur'

  type Unit =
    | 'ch'
    | 'em'
    | 'ex'
    | 'rem'
    | 'vh'
    | 'vw'
    | 'vmin'
    | 'vmax'
    | 'px'
    | 'cm'
    | 'mm'
    | 'in'
    | 'pc'
    | 'pt'
    | 'mozmm'

  export default function unit(
    unit?: Unit,
    unitPerProperty?: { [P in keyof Style]: Unit },
    isUnitlessProperty?: (property: string) => boolean
  ): (style: Style) => Style
}

declare module 'fela-plugin-validator' {
  import { Style } from 'brandeur'

  type Config = {
    logInvalid?: boolean
    deleteInvalid?: boolean
    useCSSLint?: boolean
  }
  export default function validator(config?: Config): (style: Style) => Style
}
