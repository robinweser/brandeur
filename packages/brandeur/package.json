{
  "name": "brandeur",
  "version": "0.0.5",
  "description": "Convenience layer and tool belt for css-hooks",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "sideEffects": false,
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**"
  ],
  "scripts": {
    "setup": "pnpm build",
    "clean": "rimraf lib es coverage",
    "build": "pnpm clean && pnpm build:es && pnpm build:lib",
    "build:es": "babel src --out-dir es --ignore **/__tests__",
    "build:lib": "cross-env BABEL_ENV=commonjs babel src --out-dir lib --ignore **/__tests__",
    "watch": "npm run watch:lib & npm run watch:es",
    "watch:lib": "BABEL_ENV=commonjs babel src -d lib -w",
    "watch:es": "babel src -d es -w",
    "test": "jest --config=jest.config.js",
    "coverage": "pnpm test -- --coverage"
  },
  "repository": "https://github.com/robinweser/brandeur/",
  "keywords": [
    "brandeur",
    "dynamic styling",
    "inline style",
    "css-hooks",
    "css",
    "styling",
    "functional styling",
    "functional css",
    "state-driven",
    "state-driven styling",
    "cssinjs"
  ],
  "author": "Robin Weser",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^3.1.0",
    "fast-loops": "^1.1.3",
    "isobject": "^4.0.0"
  },
  "peerDependencies": {
    "@css-hooks/react": "^1.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.23.4",
    "@babel/core": "^7.23.7",
    "@babel/node": "^7.22.19",
    "@babel/plugin-transform-modules-commonjs": "^7.23.2",
    "@babel/polyfill": "^7.12.1",
    "@babel/preset-env": "^7.23.8",
    "@css-hooks/react": "^0.0.5",
    "babel-core": "7.0.0-bridge.0",
    "babel-jest": "^29.7.0",
    "cross-env": "^7.0.3",
    "cssbeautify": "^0.3.1",
    "jest": "^29.7.0",
    "rimraf": "^5.0.5"
  }
}
