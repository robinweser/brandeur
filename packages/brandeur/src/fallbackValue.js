export default function fallbackValue(property, values, match) {
  const actualMatch = match || values[values.length - 1]

  return {
    property,
    values,
    match: actualMatch,
  }
}
