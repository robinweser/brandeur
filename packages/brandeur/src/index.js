export { default as createHooks } from './createHooks'
export { default as fallbackValue } from './fallbackValue'
