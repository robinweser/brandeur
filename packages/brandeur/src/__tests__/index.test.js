import beautify from 'cssbeautify'
import prefixer from 'brandeur-plugin-prefixer'

import brandeur from '../index'
import fallbackValue from '../fallbackValue'

describe('Using brandeur', () => {
  it('should work', () => {
    const [staticCSS, css] = brandeur({
      hooks: {
        ':hover': ':hover',
      },
      plugins: [prefixer()],
      fallbacks: [fallbackValue('position', ['-webkit-sticky', 'sticky'])],
    })

    expect(beautify(staticCSS)).toMatchSnapshot()
    expect(
      css({
        color: 'red',
        appearance: 'none',
        position: 'sticky',
        ':hover': {
          appearance: 'button',
          position: 'absolute',
        },
      })
    ).toMatchSnapshot()
  })
})
