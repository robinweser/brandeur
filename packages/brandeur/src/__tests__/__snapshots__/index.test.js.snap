// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using brandeur should work 1`] = `
":root {
    --position-sticky: ;
}

@supports (position:-webkit-sticky) {
    :root {
        --position-sticky: -webkit-sticky
    }
}

@supports (position:sticky) {
    :root {
        --position-sticky: sticky
    }
}

* {
    --zh6im6-0: initial;
    --zh6im6-1: ;
}

:hover {
    --zh6im6-0: ;--zh6im6-1:initial;
}"
`;

exports[`Using brandeur should work 2`] = `
{
  "WebkitAppearance": "var(--zh6im6-1, button) var(--zh6im6-0, none)",
  "appearance": "var(--zh6im6-1, button) var(--zh6im6-0, none)",
  "color": "red",
  "position": "var(--zh6im6-1, absolute) var(--zh6im6-0, var(--position-sticky))",
}
`;
